### PR TITLE
3box integration 2.0

### PR DESCRIPTION
**Pull request by danjm**
This PR adds on to the current 3box-integration PR. Changes include:

- switch to use spaces api instead of profile api
- ensures that state is only restored on import, and only after the user confirms via a modal prompt

e2e tests are in development, but will be added in a future PR

Demo here: https://streamable.com/ovvbf

### CODE REVIEW

**Comment by bdresser** [comment]
This looks great! Small things:
- I think we need to update the button styles to mtach the design system:

Screen Shot 2019-08-07 at 3 11 51 PM

Screen Shot 2019-08-07 at 3 12 27 PM

- Could primary button say "Restore"

**Comment by danjm** [comment]
@bdresser 

I made the threebox modal consistent with other modals while updating the wording to restore:

Screenshot from 2019-08-13 01-46-01
Screenshot from 2019-08-13 01-20-44
Screenshot from 2019-08-13 01-20-25

All of these modals have different buttons from those that you shared a screen shot of above. If the design system specifies a change to these modals design, we can do them all at once in a separate issue.

**Comment by bdresser** [comment]
cool, filed [here](https://github.com/MetaMask/metamask-extension/issues/7003), can handle later.

where did the `?` come from in the first line? let's cut it please

Screen Shot 2019-08-13 at 9 04 32 AM

**Comment by danjm** [comment]
> where did the ? come from in the first line? let's cut it please

Cut with 8363564f9

I my first job a senior engineer told me that copy+paste is the devil. Its times like these I believe it...

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [inline app/scripts/controllers/threebox.js]: What is this property for? It looks as though it's being ignored in `createMetamaskMiddleware`

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [inline app/scripts/controllers/threebox.js]: What would happen if the network was interrupted during this sync I wonder 🤔 

I thought at first that maybe there was an error not being handled somewhere, but... their API doesn't seem to handle this scenario. I guess the only option is to wait, and handle the case where it times out? 🤷‍♂

File: test/setup.js
Comment (Gudahtt) [inline test/setup.js]: It looks like this was accidentally added twice

File: ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js
Comment (Gudahtt) [inline ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: Does this need to be here at all? `restoreFromThreeBox` should already set this to true.

File: ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js
Comment (Gudahtt) [inline ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: I'm not sure whether this is needed either. Isn't it impossible for `restoreFromThreeBox` to succeed unless 3box was already initialized, and doesn't it being initialized imply that 3box was already allowed?

File: ui/app/pages/routes/index.js
Comment (Gudahtt) [inline ui/app/pages/routes/index.js]: I'm not clear on what purpose `restoredFromThreeBox` serves. It looks like the primary purpose is to prevent showing the restore confirm modal more than once in a session. However, the only case where `threeBoxSynced` could conceivably be switching from false to true is for the initial sync, unless 3box was toggled off then on again.

Is there any harm in showing the modal again after 3Box is re-enabled though? It might be useful to give the user some feedback about when syncing has finished. I'm not sure.

File: ui/app/pages/routes/index.js
Comment (Gudahtt) [inline ui/app/pages/routes/index.js]: Similar to the other case, I don't see how this could reach this state without this permission being `true` already.

File: ui/app/pages/routes/index.js
Comment (Gudahtt) [inline ui/app/pages/routes/index.js]: It looks as though this prop was added accidentally.

File: ui/app/ducks/app/app.js
Comment (Gudahtt) [inline ui/app/ducks/app/app.js]: These two actions look like they may have been added accidentally in resolving a merge conflict.

File: ui/app/store/actions.js
Comment (Gudahtt) [inline ui/app/store/actions.js]: It looks like this was added accidentally.

File: ui/app/ducks/app/app.js
Comment (Gudahtt) [inline ui/app/ducks/app/app.js]: This action appears to not exist

File: ui/app/pages/first-time-flow/create-password/import-with-seed-phrase/import-with-seed-phrase.component.js
Comment (Gudahtt) [inline ui/app/pages/first-time-flow/create-password/import-with-seed-phrase/import-with-seed-phrase.component.js]: Maybe this would be better done in `metamask-controller`? I found it a bit hard to keep track of where 3Box was being initialized in each of the three scenarios: unlocking, creating a new account, and restoring an account. The first two cases are in the controller, only the last is in the UI.

File: ui/app/pages/home/home.component.js
Comment (Gudahtt) [inline ui/app/pages/home/home.component.js]: These props don't appear to be used anymore

File: ui/app/store/actions.js
Comment (Gudahtt) [reply ui/app/store/actions.js]: Oh wait... nevermind. This is needed - it was just missing. That's concerning.

File: ui/app/ducks/app/app.js
Comment (Gudahtt) [reply ui/app/ducks/app/app.js]: Hmm.... actually, `HIDE_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` does appear to be used, as I just discovered.

`SHOW_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` is still unused though.

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [inline app/scripts/controllers/threebox.js]: This can be omitted - `updateState` updates top-level properties, leaving the rest of the state unchanged. This pattern of including a spread of the existing object is only needed for updating non-top-level things.

File: app/scripts/controllers/threebox.js
Comment (danjm) [reply app/scripts/controllers/threebox.js]: That's a good point. I guess, in the case of a timeout, we should just try again?

File: ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js
Comment (danjm) [reply ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: Your right, `setRestoredFromThreeBox(true)` can be removed

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [reply app/scripts/controllers/threebox.js]: Sure, that could work. Or just throw an error. Or maybe retry, then throw an error after X failures.

File: app/scripts/controllers/threebox.js
Comment (danjm) [reply app/scripts/controllers/threebox.js]: Yep, it can be removed

File: ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js
Comment (danjm) [reply ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: Hmm... that looks to be the case as well. What we do need here is to `_registerUpdates()`

Comment (danjm) [reply ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: which is wrapped with the "public" method `turnThreeBoxSyncingOn()`

File: app/scripts/controllers/threebox.js
Comment (danjm) [reply app/scripts/controllers/threebox.js]: - [ ] do something about this

File: ui/app/store/actions.js
Comment (Gudahtt) [inline ui/app/store/actions.js]: `setThreeBoxSyncing` doesn't appear to exist. I encountered an error for this locally by toggling the 3Box setting.

**Review by Gudahtt — DISMISSED** [review]

File: ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js
Comment (Gudahtt) [reply ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js]: Ah I see, makes sense.

File: ui/app/store/actions.js
Comment (Gudahtt) [reply ui/app/store/actions.js]: Similarly, the state `metamask.threeboxSyncing` doesn't exist either

**Comment by Gudahtt** [comment]
Are we sure that this won't ever overwrite data? It does sync on unlock, which seems like it could wipe out local state. There's no warning or anything about ti either.

I thought the plan to avoid overwriting data was to only sync on import? Maybe I'm misremembering.

File: ui/app/ducks/app/app.js
Comment (danjm) [reply ui/app/ducks/app/app.js]: No, you were right, both were accidentally re-added with a merge conflict. Here is where they were deliberately removed on a PR that was merged to develop: https://github.com/MetaMask/metamask-extension/pull/6961/commits/58f90d734ba885a2dbdd14f3c0761d0951c6cfd5

`HIDE_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` is dispatched by a method that is never actually called. I will deal with that dead code in a separate PR.

File: ui/app/ducks/app/app.js
Comment (danjm) [reply ui/app/ducks/app/app.js]: another good catch

File: ui/app/pages/home/home.component.js
Comment (danjm) [reply ui/app/pages/home/home.component.js]: :+1:

File: ui/app/pages/routes/index.js
Comment (danjm) [reply ui/app/pages/routes/index.js]: The reason that we don't want to show the modal after 3Box is re-enabled is that we want to prevent actual restoring/overwriting of local data with remote data at all times other then after import. This is to prevent a case such as: local state is A -> threebox is turned off -> local state changes to B -> threebox is turned on -> modal is shown and confirm -> local state is set back to A

We could, however, show a different modal when it is re-enabled / toggled on. Probably not necessary at this time.

Comment (danjm) [reply ui/app/pages/routes/index.js]: Moreover, `restoredFromThreeBox` persists across sessions, whereas `threeBoxSynced` defaults to false for every new session. So `restoredFromThreeBox` tells us whether the user did in fact approval the overwriting of local data with remote data after an import. This ends up preventing the modal from showing up more than "once per import".

File: ui/app/pages/routes/index.js
Comment (danjm) [reply ui/app/pages/routes/index.js]: You are right again, and same as before this needs to be replaced with `turnThreeBoxSyncingOn()`

File: ui/app/store/actions.js
Comment (danjm) [reply ui/app/store/actions.js]: In fact, it is not needed. The case where it appears in the code is never called. I've removed the addition of this via a rebase, and will remove the case where it is called in a separate PR.

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [reply app/scripts/controllers/threebox.js]: After looking over this with @whymarrh , we discovered that the `onSyncDone` function was recently updated to return a promise that correctly handles errors and resolves when the sync is finished. We'll need to update 3Box, but after that we can just await `this.box.onSyncDone()` and it'll "just work". Mostly (the timeout case is still problematic).

The "onSyncDone" from the spaces API is kinda similar - we can await `this.space.syncDone` to achieve the same affect, which is a public but undocumented property. Not ideal, but better than the current solution.

Regarding the timeout, handling that case and throwing an error is simple enough, but we're kinda stuck once the timeout happens. We can't abort the sync, so we can't reliably retry the operation. We'll have to just disable 3Box completely, and not allow the user to enable it (unless they uninstall and reinstall the extension).

If it _does_ return after the timeout, we could allow the user to retry/re-enable it. But if it times out and never returns, we're out of luck.

Comment (Gudahtt) [reply app/scripts/controllers/threebox.js]: So it looks like the changes to `onSyncDone` to return a promise haven't been released yet. I don't think they need to block the release though, as long as we handle the timeout case. The error will just get caught in handling the timeout. Far from ideal, as we can't recover from it and we force the user to wait, but we can improve things whenever those changes are released.

File: ui/app/pages/routes/index.js
Comment (Gudahtt) [reply ui/app/pages/routes/index.js]: This all makes sense; I somehow got confused about what happened upon confirmation.

**Comment by Gudahtt** [comment]
I've added a timeout for the initial sync with 3Box, after which 3Box remains disabled until the sync finishes (if ever). Here's what the settings page toggle looks like when 3Box is disabled:

3box_disabled

Really not great, as a user probably won't know what "initial sync" means, nor would they know how to fix this. But we're in an awkward spot, as there isn't any obvious way to fix the problem without... waiting, and hoping it fixes itself, or reinstalling the extension.

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [reply app/scripts/controllers/threebox.js]: Resolved for now with [47e63ed](https://github.com/MetaMask/metamask-extension/pull/6972/commits/47e63edf9ceedafa7f6f0b2a432d032297aeef7d). Explained here: https://github.com/MetaMask/metamask-extension/pull/6972#issuecomment-522718296

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [inline app/scripts/controllers/threebox.js]: Shouldn't this be prefixed with `https://`?

Comment (danjm) [reply app/scripts/controllers/threebox.js]: maybe, why do you say so?

File: app/_locales/en/messages.json
Comment (Gudahtt) [inline app/_locales/en/messages.json]: This message appears to be unused

File: app/scripts/controllers/threebox.js
Comment (Gudahtt) [reply app/scripts/controllers/threebox.js]: Without the scheme, there's no way to distinguish it from other protocols, so we effectively disallow other protocols.

The scheme was included in @danfinlay 's description of this feature in the EIP thread: https://ethereum-magicians.org/t/eip-erc-app-keys-application-specific-wallet-accounts/2742/3

Comment (danfinlay) [reply app/scripts/controllers/threebox.js]: Yeah, I agree it should have a prefix of some sort. Strangely this is a decent opportunity for a non-existent prefix, to ensure no DNS attack could impersonate this domain. For example: `browser-extension://3box.metamask.io`.

Or, maybe it would be crazy, but we could even pick a domain that is friendly towards portability between wallets. Something like `wallet://backup` or something...

File: gulpfile.js
Comment (Gudahtt) [inline gulpfile.js]: Interesting - so this task is only used for `test:copy`, which is only used by `dev:test`, which itself is only called from the `start:test` npm script.

What is the `start:test` script used for anyway? Why does it need the `localhost/8889` permission? It looks like it's basically the same as the dev build except that `IN_TEST` is set to true. I'm not sure why it'd need a special permission.

File: gulpfile.js
Comment (Gudahtt) [inline gulpfile.js]:
```suggestion
      json.permissions = [...json.permissions, 'webRequestBlocking', 'http://localhost:8889']
```
I'm guessing that `8889` is a port, not a path?

Comment (danjm) [reply gulpfile.js]: So, I don't understand why, but `http://localhost/8889` and `http://localhost/*` work, but `http://localhost:8889`, `http://localhost:*` and `http://localhost` do not

File: gulpfile.js
Comment (danjm) [reply gulpfile.js]: `start:test` shares `json.permissions = [...json.permissions, 'webRequestBlocking', 'http://localhost/8889']` with the `manifest:testing` task, which is run with `build:test`. Whenever we are running tests, we need these permissions. `'webRequestBlocking'` is needed for tests so that we can prevent certain network requests. `'http://localhost/8889'` is being added with this PR so that we can make certain http requests to a local server.

Like `start`, `start:test` intentionally excludes `bg-libs.js` from the background files listed in the manifest... because `start` doesn't build that file (I can't remember why that decision was made).

`start:test` itself is useful when you want a watched build of metamask running while working on e2e tests, or if you want to run metamask in the browser with `IN_TEST` set to true

Comment (Gudahtt) [reply gulpfile.js]: Right - I had thought at the time that the `localhost` permission was only added in this task - I hadn't realized it was in `manifest:testing` as well. I had thought that was the reason for creating `manifest:testing-local` in the first place.

But rather it looks like the reason was so that `bg-libs` could be excluded.

This seems fine, though I remain of the opinion that we'd be better off splitting those background libraries out in development and testing just as we do in production. Particularly as the gulpfile grows more complex to accommodate this.

File: gulpfile.js
Comment (Gudahtt) [reply gulpfile.js]: Interesting - apparently the match pattern here doesn't support ports?
https://bugzilla.mozilla.org/show_bug.cgi?id=1362809

I think we're better off omitting the port rather than including this non-existent "8999" path. I'm also curious whether this works on Firefox

File: ui/app/store/actions.js
Comment (Gudahtt) [reply ui/app/store/actions.js]: It looks like this is still here - did it get re-added?

File: test/e2e/mock-3box/server.js
Comment (Gudahtt) [inline test/e2e/mock-3box/server.js]: Probably no need to keep this - I assume it was just used during testing.

File: test/e2e/run-all.sh
Comment (Gudahtt) [inline test/e2e/run-all.sh]: Interesting that the end-of-file newline is missing. Our `.editorconfig` file does ask for one - maybe your editor doesn't have the editorconfig plugin installed?

File: test/e2e/threebox.spec.js
Comment (Gudahtt) [inline test/e2e/threebox.spec.js]: Another console.log from testing I assume

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: Maybe this should always grab the first address, rather than accepting an address as an input. Otherwise it could end up being initialized with different accounts depending on where in the flow it became initialized.

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: This is a bit of a nitpick, but maybe this method should be changed to no longer accept a `bool` parameter. It is never and should never be called with a `true` parameter, because that happens as part of `restoreFromThreeBox`.

Maybe it could be called `setNotRestoredFromThreeBox` - sounds kinda dumb, but I can't think of a better sounding suggestion at the moment.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

File: app/scripts/metamask-controller.js
Comment (danjm) [reply app/scripts/metamask-controller.js]: done in 40a3c5879

File: test/e2e/threebox.spec.js
Comment (danjm) [reply test/e2e/threebox.spec.js]: fixed in acd4f9bc7

File: test/e2e/run-all.sh
Comment (danjm) [reply test/e2e/run-all.sh]: fixed in acd4f9bc7

File: test/e2e/mock-3box/server.js
Comment (danjm) [reply test/e2e/mock-3box/server.js]: fixed in acd4f9bc7

File: ui/app/pages/home/home.component.js
Comment (danjm) [reply ui/app/pages/home/home.component.js]: moved to redux in aaf0f5ba3

File: ui/app/pages/home/home.component.js
Comment (danjm) [reply ui/app/pages/home/home.component.js]: fixed as part of 40a3c5879

File: app/scripts/metamask-controller.js
Comment (danjm) [reply app/scripts/metamask-controller.js]: Addressed in 37f6dfb11

File: ui/app/store/actions.js
Comment (Gudahtt) [reply ui/app/store/actions.js]: It looks like this should be here - it's just broken on develop right now :/

File: ui/app/pages/home/home.container.js
Comment (Gudahtt) [inline ui/app/pages/home/home.container.js]: It is a bit awkward that we're storing the lastUpdated time of 3Box in two places - the app state and the metamask state. Moreover, the app state last updated time is only updated once per session, then never kept up-to-date after that.

Granted, it does still serve its purpose perfectly well, but it does seem confusing.

Could we just use `metamask.lastUpdated` directly, get rid of the getter method, and get rid of the app state?

File: ui/app/store/actions.js
Comment (Gudahtt) [inline ui/app/store/actions.js]: Strictly speaking I don't think we need this change anymore, but it doesn't seem like it does any harm.

**Review by Gudahtt — CHANGES_REQUESTED** [review]
Given that we're now intending to release this as opt-in, we'd might as well make that change here before we merge this.

Maybe we could also merge this into develop sooner rather than later, and hide the option to enable 3Box behind an environment variable? Like a non-user facing feature flag. I don't think we have anything like that currently, but it'd save you the work of keeping this branch up-to-date these next couple weeks.

**Comment by Gudahtt** [comment]
Just adding a note here that #7141 conflicts with this PR in a way that git won't detect. It made a change to `development/metamaskbot-build-announce.js` that included a hard-coded list of JavaScript bundles, so that list needs to be updated to rename `lib.js` to `ui-libs.js` and to include `bg-libs.js`

File: app/scripts/migrations/037.js
Comment (Gudahtt) [inline app/scripts/migrations/037.js]: This comment should be updated

File: gulpfile.js
Comment (Gudahtt) [reply gulpfile.js]: Thoughts on whether we should change this to `localhost/*`? I remain confused about why `localhost/8999` works in the first place. If you feel inclined to leave it as `localhost/8999` we should at least make sure it works on Firefox, as I can't see any documentation to explain why it _does_ work, so it could be some sort of fluke.

File: ui/app/store/actions.js
Comment (Gudahtt) [inline ui/app/store/actions.js]: Leftover `console.log` statements here

File: ui/app/pages/home/home.component.js
Comment (Gudahtt) [inline ui/app/pages/home/home.component.js]: Conditions like this might leave us in a strange state if the flag is turned on, then off again.

This situation wouldn't occur unless someone enabled it in the dev console of course, but do you think it's worth adding the flag here to prevent it anyway? It's possible an end-user might enable 3Box in the dev console then disable it, and get in a weird state like this. Maybe unlikely though - I'm OK with ignoring this case if you'd prefer. If you did want to handle it, there are a few other examples around like this as well.

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [inline app/scripts/metamask-controller.js]: I believe this should be behind the feature flag as well, as `init` is called within the controller constructor.

Or maybe the flag should be used _within_ the controller... not sure.

**Review by Gudahtt — CHANGES_REQUESTED** [review]

**Comment by Gudahtt** [comment]
Also it'd be nice to have tests for the migration. I think all of the recent migrations have been tested, so it should be pretty easy to add using a previous one as a template.

**Comment by danjm** [comment]
@Gudahtt I realized we don't actually need a migration. The feature flag will be `undefined` until it is set to true, which will have the same result as if we migrated to explicitly be `false`.

File: app/scripts/metamask-controller.js
Comment (danjm) [reply app/scripts/metamask-controller.js]: Did it in the controller here a203d5862
If we did it in the metamask controller, we would have to had some new approach for dynamically instantiating the threebox controller based on a state change.

File: ui/app/pages/home/home.component.js
Comment (danjm) [reply ui/app/pages/home/home.component.js]: Good point. Done in 4fcc18a77

File: app/scripts/metamask-controller.js
Comment (Gudahtt) [reply app/scripts/metamask-controller.js]: Right, good point. In the controller makes much more sense.

**Review by Gudahtt — APPROVED** [review]
Looks good!

**Pull request merged**
